### PR TITLE
Convert Clasp into a source package

### PR DESCRIPTION
**Pull request by dibarbet**
Converts clasp into a source package so we can make breaking changes without breaking consumers in VS.  This package is not used for cross component communication (its an internal utility package).

In order to avoid a triple insertion, we still create a normal dll for insertion into VS.  This can be removed when all consumers have updated to the source package version.  That and additional cleanup is tracked here - https://github.com/dotnet/roslyn/issues/72251

Passing RPS - https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/532565

### CODE REVIEW

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractHandlerProvider.cs [hunk 1]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.CommonLanguageServerProtocol.Framework;
 /// <summary>
 /// Manages handler discovery and distribution.
 /// </summary>
+[System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "Temporarily required for compat - https://github.com/dotnet/roslyn/issues/72251")]
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractHandlerProvider.cs:13]: Now that the source is included in MS.CA.LanguageServer.Protocol we get public API warnings.  However rather than adding these to the public API, my plan is to convert these to internal APIs (consumers of MS.CA.LanguageServer.Protocol should not be using these types).

However, changing the APIs to internal would be a breaking change for the current VS consumers (they don't have IVTs).  Instead I'll be doing a followup PR once everyone has switched over to the source package to make these internal.  So temporarily suppressing the warnings.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/xlf/LSPFrameworkResources.cs.xlf [hunk 2]
```diff
@@ -1,42 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/xlf/LSPFrameworkResources.cs.xlf:1]: These were no longer used.  the strings are in the protocol project.  Additionally there was no resx file for any of these.

File: src/EditorFeatures/Core/Microsoft.CodeAnalysis.EditorFeatures.csproj [hunk 3]
```diff
@@ -36,6 +36,9 @@
     <PackageReference Include="Microsoft.ServiceHub.Client" />
     <PackageReference Include="Microsoft.VisualStudio.Debugger.Contracts" />
   </ItemGroup>
+  <ItemGroup>
+    <Compile Include="..\..\Compilers\Core\Portable\InternalUtilities\IsExternalInit.cs" Link="Utilities\IsExternalInit.cs" />
```
Comment (dibarbet) [inline src/EditorFeatures/Core/Microsoft.CodeAnalysis.EditorFeatures.csproj:40]: seems like a bunch of projects were picking up this from the clasp project.  Had to add it in a few places now that its become a source package.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractHandlerProvider.cs [hunk 4]
```diff
@@ -10,7 +10,11 @@ namespace Microsoft.CommonLanguageServerProtocol.Framework;
 /// <summary>
 /// Manages handler discovery and distribution.
 /// </summary>
+#if BINARY_COMPAT // TODO - Remove with https://github.com/dotnet/roslyn/issues/72251
 public abstract class AbstractHandlerProvider
+#else
+internal abstract class AbstractHandlerProvider
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractHandlerProvider.cs:16]: The source package versions of these types need to be internal.  Otherwise someone referencing Clasp directly, and MS.CA.LanguageServer.Protocol (for example, Razor) will get conflicting type definitions if they're public.

However, for binary compatibility in VS on the old non-source package versions of clasp, these APIs need to be public.  So until we move all consumers to the source only version of the package, we have to ship the dll in VS with public accessibility.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractHandlerProvider.cs [hunk 5]
```diff
@@ -2,6 +2,9 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+// This is consumed as 'generated' code in a source package and therefore requires an explicit nullable enable
+#nullable enable
```
Comment (dibarbet) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractHandlerProvider.cs:6]: we ship an .editorconfig in the package that marks all the code in here as generated code.  This ensures that analyzers do not report diagnostics for these files when included as a source package (compiler warnings are still reported - unfortunately there is no way to disable that).

And generated code requires explicit nullable defines in order to use nullable annotations.

File: src/Tools/ExternalAccess/Xaml/External/IXamlRequestHandler.cs [hunk 6]
```diff
@@ -2,11 +2,14 @@
...
 // See the LICENSE file in the project root for more information.
 
-using Microsoft.CommonLanguageServerProtocol.Framework;
+using System.Threading;
+using System.Threading.Tasks;
 
 namespace Microsoft.CodeAnalysis.ExternalAccess.Xaml;
 
 internal interface IXamlRequestHandler<TRequest, TResponse>
-    : IRequestHandler<TRequest, TResponse, XamlRequestContext>
```
Comment (dibarbet) [inline src/Tools/ExternalAccess/Xaml/External/IXamlRequestHandler.cs:10]: This was incorrectly exposing the clasp type directly to XAML.  making it internal broke their handlers.  However XAML already wrapped this handler into a Roslyn typed handler (XamlRequestHandlerBase), so all we need to do here is copy the old signature directly here and remove the inheritance.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    <Compile Include="$(MSBuildThisFileDirectory)LanguageServerEndpointAttribute.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)QueueItem.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)RequestExecutionQueue.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)RequestHandlerMetadata.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)RequestShutdownEventArgs.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)WrappedHandlerProvider.cs" />
+    <Compile Include="..\..\..\Compilers\Core\Portable\InternalUtilities\IsExternalInit.cs" Link="Utilities\IsExternalInit.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Content Include="$(MSBuildThisFileDirectory)BannedSymbols.txt" />
```
Comment (davidwengier) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems:43]: Is this file needed? It's contents are very Roslyn specific, and I would imagine the consuming projects would be the place to put it. Not that I expect it will cause any harm either 🤷‍♂️

**Review by davidwengier — APPROVED** [review]

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Package.csproj [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    <!-- Remove once https://github.com/NuGet/Home/issues/8583 is fixed -->
+    <NoWarn>$(NoWarn);NU5128</NoWarn>
+    <!-- Required to avoid NuGet ignoring the .editorconfig file -->
+    <NoDefaultExcludes>true</NoDefaultExcludes>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="StreamJsonRpc" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="..\..\..\Compilers\Core\Portable\InternalUtilities\IsExternalInit.cs" Link="Utilities\IsExternalInit.cs" />
```
Comment (ToddGrun) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Package.csproj:29]: ><Compile Include="..\..\..\Compilers\Core\Portable\InternalUtilities\IsExternalInit.cs" Link="Utilities\IsExternalInit.cs" /> [](http://example.com/codeflow?start=4&length=125)

Consuming the ISExternalInit package still isn't an option, I assume?

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  <PropertyGroup>
+    <MSBuildAllProjects Condition="'$(MSBuildVersion)' == '' Or '$(MSBuildVersion)' &lt; '16.0'">$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
+    <HasSharedItems>true</HasSharedItems>
+    <SharedGUID>64eaded3-4b5d-4431-bbe5-a4aba1c38c00</SharedGUID>
+  </PropertyGroup>
+  <PropertyGroup Label="Configuration">
+    <Import_RootNamespace>Microsoft.CommonLanguageServerProtocol.Framework.Shared</Import_RootNamespace>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="$(MSBuildThisFileDirectory)AbstractHandlerProvider.cs" />
```
Comment (ToddGrun) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems:12]: >AbstractHandlerProvider.cs [](http://example.com/codeflow?start=49&length=26)

Out of curiosity, why not an all-inclusive glob?

Comment (dibarbet) [reply src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems:12]: Was autogenerated by VS - also seems to match what other shared projects do (probably because they were also autogenerated)
https://github.com/dotnet/roslyn/blob/main/src/Dependencies/PooledObjects/Microsoft.CodeAnalysis.PooledObjects.projitems

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems (hunk 7 above)
Comment (ToddGrun) [reply src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems:43]: Seems a bit weird to me too

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Package.csproj (hunk 8 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Package.csproj:29]: this one?  https://www.nuget.org/packages/IsExternalInit. I guess probably possible, but we don't own it and its one less dependency to manage, so I'd rather just use our version.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems (hunk 7 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Microsoft.CommonLanguageServerProtocol.Framework.Shared.projitems:43]: yeah, I'll remove this.  Shouldn't be necessary here.

**Review by ToddGrun — APPROVED** [review]
:shipit:
